### PR TITLE
Add a script to generate unreleased changelog

### PR DESCRIPTION
**Pull request by sosukesuzuki**
## Description

I found it tedious to manually create a changelog each time. What do you think about adding a generating script?

Pass the pull request number and category as arguments.
Get the PR title and author name via the GitHub API and create a file in `./changelog_unreleased`.

````
$ node ./scripts/generate-changelog.mjs 10961 typescript

$ cat ./changelog_unreleased/typescript/10961.md
<!--

1. Choose a folder based on which language your PR is for.

   - For JavaScript, choose `javascript/` etc.
   - For TypeScript specific syntax, choose `typescript/`.
   - If your PR applies to multiple languages, such as TypeScript/Flow, choose one folder and mention which languages it applies to.

2. In your chosen folder, create a file with your PR number: `XXXX.md`. For example: `typescript/6728.md`.

3. Copy the content below and paste it in your new file.

4. Fill in a title, the PR number and your user name.

5. Optionally write a description. Many times it’s enough with just sample code.

6. Change ```jsx to your language. For example, ```yaml.

7. Change the `// Input` and `// Prettier` comments to the comment syntax of your language. For example, `# Input`.

8. Choose some nice input example code. Paste it along with the output before and after your PR.

-->

#### Fix order of override modifiers (#10961 by @sosukesuzuki)

<!-- Optional description if it makes sense. -->

<!-- prettier-ignore -->
```jsx
// Input
(foo ?? baz) || baz;

// Prettier stable
foo ?? baz || baz;

// Prettier main
(foo ?? baz) || baz;
```
````

If accepted, I will mention about this script in `CONTRIBUTING.md`.

## Checklist

- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

File: .prettierrc [hunk 1]
```diff
@@ -1,4 +1,4 @@
 overrides:
-  - files: "**/*.{js,mjs}"
+  - files: "**/*.{js}"
```
Comment (sosukesuzuki) [inline .prettierrc:2]: This will be removed when meriyah supports Top-Level await

File: scripts/generate-changelog.mjs [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+  const changelogTemplate = await fs.readFile(changelogTemplatePath, "utf-8");
+
+  const titlePart = "Title";
+  const prNumberPart = "#XXXX";
+  const userPart = "@user";
+
+  return changelogTemplate
+    .replace(titlePart, title)
+    .replace(prNumberPart, `#${prNumber}`)
+    .replace(userPart, `@${user}`);
```
Comment (fisker) [inline scripts/generate-changelog.mjs:92]: Can we replace language of code block, and comment style?

File: scripts/generate-changelog.mjs [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
+  return changelogTemplate
+    .replace(titlePart, title)
+    .replace(prNumberPart, `#${prNumber}`)
+    .replace(userPart, `@${user}`)
+    .replace(codeBlockPart, `\`\`\`${syntax}\n`)
+    .replace(inputCommentPart, getCommentForSyntax(syntax, "Input") + "\n")
+    .replace(
+      stableCommentPart,
+      getCommentForSyntax(syntax, "Prettier stable") + "\n"
```
Comment (fisker) [inline scripts/generate-changelog.mjs:104]: Do you think we can also auto generate stable and main result by `require('prettier').format` and `require('.').format` ?

Comment (sosukesuzuki) [reply scripts/generate-changelog.mjs:104]: I think it's possible, but it's hard to think of a good interface. I think at least it isn't in scope of this PR.

File: .eslintignore [hunk 4]
```diff
@@ -11,3 +11,7 @@
 /website/build/
 /website/static/playground.js
 /website/static/lib/
+
+# This is temporal, will be removed when ESLint supports Top-Level await.
+# https://github.com/eslint/eslint/issues/14632
+/scripts/generate-changelog.mjs
```
Comment (fisker) [inline .eslintignore:17]: How about not use top-level await for now? We can use [`unicorn/prefer-top-level-await`](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/prefer-top-level-await.md) when ESLint is ready.

File: scripts/generate-changelog.mjs [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+
+import fs from "node:fs/promises";
+import path from "node:path";
+import fetch from "node-fetch";
+import createEsmUtils from "esm-utils";
+import { CHANGELOG_CATEGORIES } from "./utils/changelog-categories.mjs";
+
+const { __dirname } = createEsmUtils(import.meta);
+
+const [prNumberString, category] = process.argv.slice(2);
```
Comment (fisker) [inline scripts/generate-changelog.mjs:17]: Can we get those by select and input? Maybe by [`Inquirer.js`](https://github.com/SBoudrias/Inquirer.js), [`enquirer`](https://github.com/enquirer/enquirer), or whatever you prefer.

File: .eslintignore (hunk 4 above)
Comment (sosukesuzuki) [reply .eslintignore:17]: Sounds good. It is a good ESLint rule.

File: scripts/generate-changelog.mjs (hunk 5 above)
Comment (sosukesuzuki) [reply scripts/generate-changelog.mjs:17]: It is helpful especially for beginners. Thanks.

File: scripts/generate-changelog.mjs [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+(async () => {
+  const prNumberPrompt = new enquirer.NumberPrompt({
+    message: "Input your Pull Request number:",
+  });
+  const prNumberString = Number.parseInt(await prNumberPrompt.run(), 10);
+
+  const categoryPrompt = new enquirer.AutoComplete({
+    message: "Input category of your Pull Request:",
+    limit: CHANGELOG_CATEGORIES.length,
+    choices: [...CHANGELOG_CATEGORIES],
```
[the remark this answers is not in the document]
Comment (sosukesuzuki) [reply scripts/generate-changelog.mjs:27]: Yes, for some reason, it destroys the choices array.

[the remark this answers is not in the document]
Comment (fisker) [reply scripts/generate-changelog.mjs:27]: Maybe add a comment?

File: scripts/generate-changelog.mjs [hunk 7]
```diff
@@ -0,0 +1,188 @@
+#!/usr/bin/env node
+
+/**
+ * Pass the pull request number and category as arguments.
+ * Get the PR title and author name via the GitHub API and create a file in ./changelog_unreleased
+ *   $ node ./scripts/generate-changelog.mjs 10961 typescript
+ */
```
Comment (fisker) [inline scripts/generate-changelog.mjs:7]: please update this command.

**Comment by fisker** [comment]
We  should mention this script in https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md#other

**Review by fisker — APPROVED** [review]

**Pull request merged**